admin/services/data-access/pupil-census-import.data.service.js

```diff
@@ -0,0 +1,72 @@
+'use strict'
+
+const { TYPES } = require('tedious')
+const moment = require('moment')
+
+const config = require('../../config')
+const pupilCensusImportDataService = {}
+const sqlService = require('./sql.service')
+const sqlPoolService = require('./sql.pool.service')
+
+/**
+ * Execute pupil data bulk import
+ * @param {Array} pupilData
+ * @param {Array} schools
+ * @param {Number} jobId
+ * @return {Promise<*>}
+ */
+pupilCensusImportDataService.sqlBulkImport = async(pupilData, schools, jobId) => {
+  const result = {}
+  const con = await sqlPoolService.getConnection()
+
+  try {
+    result.output = await bulkLoadData(con, pupilData, schools, jobId)
+  } catch (error) {
+    result.errorOutput = error
+  }
+  con.release()
+  return result
+}
+
+const bulkLoadData = (connection, pupilData, schools, jobId) => {
+  return new Promise((resolve, reject) => {
+    const bulkLoad = connection.newBulkLoad(`${config.Sql.Database}.${sqlService.adminSchema}.[pupil]`, function (error, rowCount) {
+      if (error) {
+        return reject(error)
+      }
+      resolve(`Inserted ${rowCount} rows`)
+    })
+
+    bulkLoad.addColumn('school_id', TYPES.Int, {nullable: false})
+    bulkLoad.addColumn('upn', TYPES.Char, {length: 13, nullable: false})
+    bulkLoad.addColumn('lastName', TYPES.NVarChar, {length: 'max', nullable: false})
+    bulkLoad.addColumn('foreName', TYPES.NVarChar, {length: 'max', nullable: false})
+    bulkLoad.addColumn('middleNames', TYPES.NVarChar, {length: 'max', nullable: true})
+    bulkLoad.addColumn('gender', TYPES.Char, {length: 1, nullable: false})
+    bulkLoad.addColumn('dateOfBirth', TYPES.DateTimeOffset, {nullable: false})
+    bulkLoad.addColumn('job_id', TYPES.Int, {nullable: true})
+
+    for (let index = 0; index < pupilData.length; index++) {
```

Comment:
I'm not an SQL Server expert and haven't studied the internals of the underlying Tedious library but it seems optimistic to expect the entire file to be inserted into single bulk load insert.  Memory and DB performance may be adversely affected.

What if someone passes in a large file of 600K records?  Can it cope?  What does the memory profile look like?

Reply:
@jon-shipley At this stage we are dealing with 6K records and bulk insertion has proven to be swift enough to handle it without any issues.

Reply:
@GeorgeChatzigiannis that's an assumption based on local machine testing.  I suggest we get this on deployed to Azure asap and test it out with 6K file upload